sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala

```diff
@@ -144,6 +146,56 @@ object EliminateSerialization extends Rule[LogicalPlan] {
 }
 
 /**
+ * Add Filter to left and right of an inner Join to filter out rows with null keys.
+ * So we may not need to check nullability of keys while joining. Besides, by filtering
+ * out keys with null, we can also reduce data size in Join.
+ */
+object AddFilterOfNullForInnerJoin extends Rule[LogicalPlan] with PredicateHelper {
```

Comment:
We could also do this for semi join and outer join, should we call it AddFilterOfNullForEquiJoin ?